File: app/src/main/java/com/worldventures/dreamtrips/wallet/ui/settings/help/video/impl/WalletHelpVideoDelegate.java
package com.worldventures.dreamtrips.wallet.ui.settings.help.video.impl;

import android.content.Context;
import android.net.Uri;

import com.innahema.collections.query.queriables.Queryable;
import com.worldventures.dreamtrips.core.utils.ProjectTextUtils;
import com.worldventures.dreamtrips.social.ui.video.model.CachedModel;
import com.worldventures.dreamtrips.social.ui.video.model.VideoLanguage;
import com.worldventures.dreamtrips.social.ui.video.model.VideoLocale;
import com.worldventures.dreamtrips.wallet.ui.settings.help.video.WalletHelpVideoScreen;
import com.worldventures.dreamtrips.wallet.ui.settings.help.video.model.WalletVideoModel;

import java.io.File;
import java.util.List;
import java.util.Locale;

public class WalletHelpVideoDelegate {

   private List<VideoLocale> videoLocales = null;
   private VideoLocale lastVideoLocale = null;

   private final Context context;

   public WalletHelpVideoDelegate(Context context) {
      this.context = context;
   }

   public VideoLanguage getDefaultLanguage(final List<VideoLocale> videoLocales) {
      VideoLocale videoLocale = null;
      if (videoLocales != null && !videoLocales.isEmpty()) {
         videoLocale = Queryable.from(videoLocales).firstOrDefault(element -> element.getCountry()
               .equalsIgnoreCase(Locale.getDefault().getCountry()));
      }

      //for retry when HttpError
      if (videoLocale == null && lastVideoLocale != null) videoLocale = lastVideoLocale;

      return getDefaultLanguage(videoLocale);
   }

   private VideoLanguage getDefaultLanguage(VideoLocale videoLocale) {
      if (videoLocale != null) {
         final VideoLanguage videoLanguage = Queryable.from(videoLocale.getLanguages())
               .firstOrDefault(language -> language.getLocaleName()
                     .equalsIgnoreCase(Locale.getDefault().getLanguage()));
         if (videoLanguage != null) return videoLanguage;
      }
      return new VideoLanguage(
            Locale.getDefault().getDisplayLanguage(),
            getDefaultLocaleName());
   }

   private String getDefaultLocaleName() {
      return Locale.getDefault().getLanguage() + "-" + Locale.getDefault().getCountry().toLowerCase();
   }

   public String getPathForCache(final CachedModel entity) {
      return getFilePath(entity.getUrl());
   }

   public String obtainVideoLanguage(final WalletVideoModel video) {
      return ProjectTextUtils.defaultIfEmpty(video.getVideo().getLanguage(), "null");
   }

   public Uri playVideo(final WalletVideoModel video) {
      CachedModel videoEntity = video.getVideo().getCacheEntity();
      Uri parse = Uri.parse(video.getVideo().getVideoUrl());
      if (isCached(videoEntity)) {
         parse = Uri.parse(getFilePath(videoEntity.getUrl()));
      }
      return parse;
   }

   public boolean isCurrentSelectedVideoLocale(final VideoLocale videoLocale) {
      return lastVideoLocale == null || videoLocale.equals(lastVideoLocale);
   }

   public void processCachingState(final CachedModel cachedEntity, final WalletHelpVideoScreen view) {
      Queryable.from(view.getCurrentItems())
            .notNulls()
            .map(WalletVideoModel::getVideo)
            .filter(video -> video.getCacheEntity().getUuid().equals(cachedEntity.getUuid()))
            .forEachR(video -> {
               video.setCacheEntity(cachedEntity);
               view.notifyItemChanged(cachedEntity);
            });
   }

   public int getDefaultLocaleIndex(List<VideoLocale> videoLocales) {
      VideoLocale videoLocale = Queryable.from(videoLocales).firstOrDefault(element -> element.getCountry()
            .equalsIgnoreCase(Locale.getDefault().getCountry()));
      return videoLocale == null ? 0 : videoLocales.indexOf(videoLocale);
   }

   public int getLastSelectedLocaleIndex() {
      return videoLocales == null ? 0 : videoLocales.indexOf(lastVideoLocale);
   }

   public void setVideoLocales(List<VideoLocale> videoLocales) {
      this.videoLocales = videoLocales;
   }

   public void setCurrentSelectedVideoLocale(VideoLocale videoLocale) {
      this.lastVideoLocale = videoLocale;
   }

   public VideoLanguage getDefaultLanguageFromLastLocales() {
      return getDefaultLanguage(lastVideoLocale);
   }

   private boolean isCached(CachedModel cachedModel) {
      return new File(getFilePath(cachedModel.getUrl())).exists() && cachedModel.getProgress() == 100;
   }

   private String getFilePath(String url) {
      return context.getFilesDir().getPath() + File.separator + getFileName(url);
   }

   private String getFileName(String url) {
      return url.substring(url.lastIndexOf("/") + 1);
   }
}

File: wallet/src/main/java/com/worldventures/wallet/ui/settings/help/video/impl/WalletHelpVideoDelegate.kt
package com.worldventures.wallet.ui.settings.help.video.impl

import android.content.Context
import android.net.Uri
import com.worldventures.core.model.CachedModel
import com.worldventures.core.modules.video.model.VideoLanguage
import com.worldventures.core.modules.video.model.VideoLocale
import com.worldventures.core.utils.ProjectTextUtils.defaultIfEmpty
import com.worldventures.wallet.ui.settings.help.video.WalletHelpVideoScreen
import com.worldventures.wallet.ui.settings.help.video.model.WalletVideoModel
import java.io.File
import java.util.Locale

private const val MAX_PROGRESS = 100

class WalletHelpVideoDelegate(private val context: Context) {

   private var videoLocales: List<VideoLocale>? = null
   private var lastVideoLocale: VideoLocale? = null

   private val defaultLocaleName: String
      get() = Locale.getDefault().language + "-" + Locale.getDefault().country.toLowerCase(Locale.US)

   val lastSelectedLocaleIndex: Int
      get() = videoLocales?.indexOf(lastVideoLocale) ?: 0

   val defaultLanguageFromLastLocales: VideoLanguage
      get() = getDefaultLanguage(lastVideoLocale)

   fun getDefaultLanguage(videoLocales: List<VideoLocale>?): VideoLanguage {
      var videoLocale: VideoLocale? = null
      if (videoLocales != null && videoLocales.isNotEmpty()) {
         videoLocale = videoLocales.firstOrNull {
            it.country.equals(Locale.getDefault().country, ignoreCase = true)
         }
      }

      //for retry when HttpError
      if (videoLocale == null && lastVideoLocale != null) {
         videoLocale = lastVideoLocale
      }

      return getDefaultLanguage(videoLocale)
   }

   private fun getDefaultLanguage(videoLocale: VideoLocale?): VideoLanguage {
      if (videoLocale != null) {
         val videoLanguage = videoLocale.languages
               .firstOrNull { it.localeName.equals(Locale.getDefault().language, ignoreCase = true) }
         if (videoLanguage != null) {
            return videoLanguage
         }
      }
      return VideoLanguage(Locale.getDefault().displayLanguage, defaultLocaleName)
   }

   fun getPathForCache(entity: CachedModel): String = getFilePath(entity.url)

   fun obtainVideoLanguage(video: WalletVideoModel): String = defaultIfEmpty(video.video.language, "null")

   fun playVideo(video: WalletVideoModel): Uri {
      val videoEntity = video.video.cacheEntity
      var parse = Uri.parse(video.video.videoUrl)
      if (isCached(videoEntity)) {
         parse = Uri.parse(getFilePath(videoEntity.url))
      }
      return parse
   }

   fun isCurrentSelectedVideoLocale(videoLocale: VideoLocale) = lastVideoLocale == null || videoLocale == lastVideoLocale

   fun processCachingState(cachedEntity: CachedModel, view: WalletHelpVideoScreen) {
      view.currentItems
            .filter { it != null }
            .map { it.video }
            .filter({ video -> video.cacheEntity.uuid == cachedEntity.uuid })
            .forEach({ video ->
               video.cacheEntity = cachedEntity
               view.notifyItemChanged(cachedEntity)
            })
   }

   fun getDefaultLocaleIndex(videoLocales: List<VideoLocale>): Int {
      val videoLocale = videoLocales.firstOrNull {
         it.country.equals(Locale.getDefault().country, ignoreCase = true)
      }
      return videoLocale?.let { videoLocales.indexOf(it) } ?: 0
   }

   fun setVideoLocales(videoLocales: List<VideoLocale>) {
      this.videoLocales = videoLocales
   }

   fun setCurrentSelectedVideoLocale(videoLocale: VideoLocale) {
      this.lastVideoLocale = videoLocale
   }

   private fun isCached(cachedModel: CachedModel): Boolean =
         File(getFilePath(cachedModel.url)).exists() && cachedModel.progress == MAX_PROGRESS

   private fun getFilePath(url: String): String = context.filesDir.path + File.separator + getFileName(url)

   private fun getFileName(url: String): String = url.substring(url.lastIndexOf('/') + 1)
}
